include/drivers/syscon.h

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2021 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public SYSCON driver APIs
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_SYSCON_H_
+#define ZEPHYR_INCLUDE_DRIVERS_SYSCON_H_
+
+/**
+ * @brief SYSCON Interface
+ * @defgroup syscon_interface SYSCON Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/types.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @typedef syscon_api_get_base()
+ * @brief Callback API for getting the base address
+ *
+ * @see syscon_get_base() for argument descriptions.
+ */
+typedef uintptr_t (*syscon_api_get_base)(const struct device *dev);
+
+/**
+ * @typedef syscon_api_read_reg()
+ * @brief Callback API for reading a syscon register
+ *
+ * @see syscon_read_reg() for argument descriptions.
+ */
+typedef int (*syscon_api_read_reg)(const struct device *dev, uint16_t reg,
+				   uint32_t *val);
+
+/**
+ * @typedef syscon_api_write_reg()
+ * @brief Callback API for writing a syscon register
+ *
+ * @see syscon_write_reg() for argument descriptions.
+ */
+typedef int (*syscon_api_write_reg)(const struct device *dev, uint16_t reg,
+				    uint32_t val);
+
+/**
+ * @brief SYSCON driver API
+ */
+__subsystem struct syscon_driver_api {
+	syscon_api_get_base get_base;
+	syscon_api_read_reg read_reg;
+	syscon_api_write_reg write_reg;
+};
+
+/**
+ * @brief Get the syscon base address
+ *
+ * This function returns the syscon base address
+ *
+ * @param dev SYSCON device
+ *
+ * @return 0 on error, the base address on success
+ */
+static inline uintptr_t syscon_get_base(const struct device *dev)
```

Comment:
Should the driver giving out the base address? Just wondering how useful it is for applications.

Reply:
Yeah so, the rationale here is that the user can interface with the syscon region in three ways:
- Using DT helpers, basically retrieving the base address through the phandle and `DT_REG_ADDR` and using that
- Using the syscon driver with `syscon_get_base` + `syscon_{read,write}_reg`
- A mixed approach, maybe retrieving the base address with `syscon_get_base` and then relying on the usual `sys_{read,write}32`

For the latest two having a `syscon_get_base` could be handy after all.

Reply:
Understood. I thought the `reg` argument to `syscon_{read,write}_reg` would be an offset to base address, so there would be no need for the outside world to know the base address.

I don't have any issue with this function.